bigtable/tests/data_integration_test.cc

```diff
@@ -211,6 +215,130 @@ TEST_F(DataIntegrationTest, TableReadRowNotExistTest) {
+  auto table = CreateTable(table_name, table_config);
+  std::string const row_key1 = "row-key-1";
+  std::string const row_key2 = "row-key-2";
+  std::string const row_key3 = "row-key-3";
+
+  std::vector<bigtable::Cell> created{
+      {row_key1, family, "c1", 1000, "data1", {}},
+      {row_key1, family, "c2", 1000, "data2", {}},
+      {row_key2, family, "c1", 1000, "data3", {}},
+      {row_key3, family, "c1", 1000, "data4", {}}};
+
+  CreateCells(*table, created);
+
+  std::vector<bigtable::Cell> expected{
+      {row_key1, family, "c1", 1000, "data1", {}},
+      {row_key1, family, "c2", 1000, "data2", {}},
+      {row_key2, family, "c1", 1000, "data3", {}}};
+
+  // Some equivalent ways of reading just the first two rows
+  auto read1 =
+      table->ReadRows(bigtable::RowSet(bigtable::RowRange::InfiniteRange()), 2,
+                      bigtable::Filter::PassAllFilter());
+  CheckEqualUnordered(expected, MoveCellsFromReader(read1));
+
+  bigtable::RowSet rs2;
+  rs2.Append(row_key1);
+  rs2.Append(row_key2);
+  auto read2 =
+      table->ReadRows(std::move(rs2), bigtable::Filter::PassAllFilter());
+  CheckEqualUnordered(expected, MoveCellsFromReader(read2));
+
+  bigtable::RowSet rs3(bigtable::RowRange::Closed(row_key1, row_key2));
+  auto read3 =
+      table->ReadRows(std::move(rs3), bigtable::Filter::PassAllFilter());
+  CheckEqualUnordered(expected, MoveCellsFromReader(read3));
+
+  DeleteTable(table_name);
+}
+
+TEST_F(DataIntegrationTest, TableReadRowsNoRows) {
+  std::string const table_name = "table-read-rows-no-rows";
+  auto table = CreateTable(table_name, table_config);
+  std::string const row_key1 = "row-key-1";
+  std::string const row_key2 = "row-key-2";
+  std::string const row_key3 = "row-key-3";
+
+  std::vector<bigtable::Cell> created{
+      {row_key1, family, "c1", 1000, "data1", {}},
+      {row_key3, family, "c1", 1000, "data2", {}}};
+
+  CreateCells(*table, created);
+
+  std::vector<bigtable::Cell> expected;  // empty
+
+  // read nonexistent rows
+  auto read1 = table->ReadRows(bigtable::RowSet(row_key2),
+                               bigtable::Filter::PassAllFilter());
+  CheckEqualUnordered(expected, MoveCellsFromReader(read1));
+
+  auto read2 =
+      table->ReadRows(bigtable::RowSet(bigtable::RowRange::Prefix(row_key2)),
+                      bigtable::Filter::PassAllFilter());
+  CheckEqualUnordered(expected, MoveCellsFromReader(read2));
+
+  // TODO(#404): also call with bigtable::RowSet(bigtable::RowRange::Empty())
+
+  DeleteTable(table_name);
+}
+
+TEST_F(DataIntegrationTest, TableReadRowsWrongTable) {
+  std::string const table_name = "table-read-rows-wrong-table";
+  bigtable::Table table(data_client_, table_name);
+
+  auto read1 =
+      table.ReadRows(bigtable::RowSet(bigtable::RowRange::InfiniteRange()),
+                     bigtable::Filter::PassAllFilter());
+#if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+  EXPECT_THROW(read1.begin(), std::runtime_error);
```

Comment:
I think I know why this fails with no exceptions enabled.  Calling `read1.begin()` is an error, but we no longer crash the application when the iterator first fails.  The application needs to call `read1.Finish()` to get the error status, and only if they fail to retrieve the status does the code crash (to avoid silent errors).

Reply:
Thanks for pointing that out. But I actually got confused by the line here https://github.com/GoogleCloudPlatform/google-cloud-cpp/blob/dc5c782d69c8f455f075ba602aa90183fbb427c5/bigtable/client/table.cc#L74
that always sets `raise_on_error` to true... and I still am, it seems to me it's still going down that code path in the `noex::` namespace?

I reran the test, with
```cpp
  EXPECT_EQ(read1.begin(), read1.end());
```
and got this:

```
Aborting because exceptions are disabled: Unretriable error: table "projects/emulated-1524060748/instances/data-test/tables/table-read-rows-wrong-table" not found
```
so the program does die.

Is the no-exception code path ready? otherwise we could submit without this test and fix it separately together with that.

Reply:
I've changed noex::Table::ReadRows to use raise_on_error=false, I think that should do the trick. PTAL